### PR TITLE
[#14] Add form for track

### LINKED ISSUE #14: Form for Track

### PR DESCRIPTION
**Pull request by Voovaaa**

### CODE REVIEW

File: first_django_project/discography/forms.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from django import forms
+from django.core.exceptions import ValidationError
+from .models import Artist
+import datetime
+
+
+class TrackForm(forms.Form):
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: use ModelForm

File: first_django_project/discography/forms.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    artist_id = forms.IntegerField(label="Artist id (you can see in URL)")
+
+
+    def clean_name(self):
+        if len(self.cleaned_data['name']) > 2:
+            return self.cleaned_data['name']
+        raise ValidationError(("Invalid name: less than 3 letters"))
+
+    def clean_date(self):
+        if (self.cleaned_data['date'].year < datetime.date.today().year
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Difficult. Use timezone.now from django.urls

File: first_django_project/discography/forms.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                or (self.cleaned_data['date'].year == datetime.date.today().year
+                and self.cleaned_data['date'].month < datetime.date.today().month)):
+            return self.cleaned_data['date']
+        raise ValidationError(("Invalid date: track can't be published in future"))
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Delete unnecessary brackets

File: first_django_project/discography/forms.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    def clean_artist_id(self):
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Delete

File: first_django_project/discography/forms.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        if self.cleaned_data["artist_id"] <= len(Artist.objects.all()) and self.cleaned_data["artist_id"] > 0:
+            return self.cleaned_data["artist_id"]
+        raise ValidationError(("Invalid artist id"))
+
+    def clean_data(self):
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Remove

File: first_django_project/discography/views.py [hunk 6]
```diff
@@ -14,3 +15,15 @@ class ArtistDetailView(generic.DetailView):
     model = Artist
 
 
+class TrackFormView(generic.FormView):
+    template_name = "discography/new_track.html"
+    form_class = forms.TrackForm
+    success_url = "/"
+
+
+    def form_valid(self, form):
```
Comment (Safintim) [inline first_django_project/discography/views.py]: Remove

File: first_django_project/discography/views.py [hunk 7]
```diff
@@ -14,3 +15,15 @@ class ArtistDetailView(generic.DetailView):
     model = Artist
 
 
+class TrackFormView(generic.FormView):
```
Comment (Safintim) [inline first_django_project/discography/views.py]: use CreateView

File: first_django_project/discography/templates/base_form.html [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{% extends "base.html" %}
```
Comment (Safintim) [inline first_django_project/discography/templates/base_form.html]: Delete

**Comment by Voovaaa** [comment]
чёт я не то с гитом сделал этим force-pushed

File: first_django_project/discography/forms.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+from django import forms
+from django.core.exceptions import ValidationError
+from .models import Artist, Track
+from django.utils import timezone
+import datetime
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Remove

File: first_django_project/discography/forms.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class TrackForm(forms.ModelForm):
+    class Meta:
+        model = Track
+        fields = ["name", "date_published", "artist"]
+
+
+    def clean_name(self):
+        if len(self.cleaned_data['name']) < 3:
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Where changes in Track model? Where validator for model field?

File: first_django_project/discography/forms.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+            raise ValidationError("Invalid name: less than 3 letters")
+        return self.cleaned_data['name']
+
+    def clean_date_published(self):
+        if self.cleaned_data['date_published'] > timezone.now().date(): # DOESNT WORK
```
Comment (Safintim) [inline first_django_project/discography/forms.py]: Remove comment

File: first_django_project/discography/templates/base_form.html [hunk 12]
```diff
@@ -0,0 +1,10 @@
+{% extends "base.html" %}
+
```
Comment (Safintim) [inline first_django_project/discography/templates/base_form.html]: Delete this file

File: first_django_project/discography/views.py [hunk 13]
```diff
@@ -1,6 +1,7 @@
 from django.shortcuts import render
 from django.views import generic
 from .models import Artist, Track
+from . import forms
```
Comment (Safintim) [inline first_django_project/discography/views.py]: import TrackForm

File: first_django_project/discography/forms.py (hunk 10 above)
Comment (Safintim) [reply first_django_project/discography/forms.py]: And migrations

File: first_django_project/discography/models.py [hunk 14]
```diff
@@ -10,10 +15,10 @@ def __str__(self):
 
 
 class Track(models.Model):
-    name = models.CharField(max_length=100)
+    name = models.CharField(max_length=100, validators=[validate_track_name])
```
Comment (Safintim) [inline first_django_project/discography/models.py]: use MinLengthValidator

File: first_django_project/discography/migrations/0003_alter_track_name.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('discography', '0002_alter_track_artist'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='track',
+            name='name',
+            field=models.CharField(max_length=100, validators=[discography.models.validate_track_name]),
```
Comment (Safintim) [inline first_django_project/discography/migrations/0003_alter_track_name.py]: Fix migration. Delete and create or change validators

Comment (Voovaaa) [reply first_django_project/discography/migrations/0003_alter_track_name.py:17]: я, не подумав, все удалил миграции

**Review by Safintim — APPROVED** [review]

**Pull request merged**